server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -83,27 +111,99 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
         for (IndexRoutingTable indexShardRouting : state.routingTable()) {
             for (int i = 0; i < indexShardRouting.size(); i++) {
                 IndexShardRoutingTable shardRouting = indexShardRouting.shard(i);
-                status.addPrimary(shardRouting.primaryShard(), shutdown);
+                status.addPrimary(shardRouting.primaryShard(), state, shutdown, includeDetails);
                 for (ShardRouting replicaShard : shardRouting.replicaShards()) {
-                    status.addReplica(replicaShard, shutdown);
+                    status.addReplica(replicaShard, state, shutdown, includeDetails);
                 }
             }
         }
 
-        return createIndicator(status.getStatus(), status.getSummary(), status.getDetails(includeDetails), status.getImpacts());
+        return new HealthIndicatorResult(
+            name(),
+            component(),
+            status.getStatus(),
+            status.getSummary(),
+            status.getDetails(includeDetails),
+            status.getImpacts(),
+            status.getUserActions(includeDetails)
+        );
     }
 
-    private static class ShardAllocationCounts {
+    // TODO: #85572 Fill in help URLs once they are finalized
+    public static final UserAction.Definition ACTION_RESTORE_FROM_SNAPSHOT = new UserAction.Definition(
+        "restore_from_snapshot",
+        Explanations.Allocation.NO_COPIES,
+        null
+    );
+    public static final UserAction.Definition ACTION_CHECK_ALLOCATION_EXPLAIN_API = new UserAction.Definition(
+        "explain_allocations",
+        "Elasticsearch isn't allowed to allocate some shards from these indices to any of the nodes in the cluster. Diagnose the issue by "
+            + "calling the allocation explain api for an index [GET _cluster/allocation/explain]. Choose a node to which you expect a "
+            + "shard to be allocated, find this node in the node-by-node explanation, and address the reasons which prevent Elasticsearch "
+            + "from allocating a shard there.",
+        null
+    );
+    public static final UserAction.Definition ACTION_ENABLE_ALLOCATIONS = new UserAction.Definition(
+        "enable_allocations",
+        "Elasticsearch isn't allowed to allocate some shards from these indices because allocation for those shards has been disabled. "
+            + "Check that the ["
+            + EnableAllocationDecider.INDEX_ROUTING_ALLOCATION_ENABLE_SETTING.getKey()
+            + "] index settings and the ["
+            + EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING.getKey()
+            + "] cluster setting are not set to ["
+            + EnableAllocationDecider.Allocation.NONE.toString().toLowerCase(Locale.getDefault())
```

Comment:
Can we change this to be something like "Check that the [...] index settings and [...] cluster settings are set to [all]"?

Also, +1 to what Andrei said below about trying to determine whether it's a cluster-level or index-level in the message.

Reply:
Applied in [bbd994d](https://github.com/elastic/elasticsearch/pull/85071/commits/bbd994dd652e9ea496ecca48a894318852a93996)